### ISSUE TITLE
Swarm service update with --reserve-memory causes panic: runtime error

### ISSUE THREAD

**Issue by jjcorrea** [moby/moby#38363]
**Description**

**Steps to reproduce the issue:**

We tested the very same scenario in two different environments running the same docker-ce version, and only one of those makes the issue occur. 

1. docker service create --name redis redis:3.0.6
2. docker service update redis --reserve-memory=128M

**Describe the results you received:**
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x55726d231d63]

goroutine 1 [running]:
github.com/docker/cli/cli/command/service.updateService(0x55726e499960, 0xc4200cc020, 0x7fe224047120, 0xc4205c2100, 0xc4204d1c00, 0xc420227960, 0xc4201f0940, 0x19)
	/go/src/github.com/docker/cli/cli/command/service/update.go:333 +0x513
github.com/docker/cli/cli/command/service.runUpdate(0x55726e4b2200, 0xc4203cdb00, 0xc4204d1c00, 0xc42028c800, 0x7ffe115786a2, 0xf, 0x55726c91bbe0, 0xc420227c48)
	/go/src/github.com/docker/cli/cli/command/service/update.go:172 +0x24e
github.com/docker/cli/cli/command/service.newUpdateCommand.func1(0xc42055a280, 0xc420586920, 0x1, 0x2, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/service/update.go:33 +0x98
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).execute(0xc42055a280, 0xc4200d0120, 0x2, 0x2, 0xc42055a280, 0xc4200d0120)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:762 +0x46a
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4203d3b80, 0xc42023dfa0, 0x55726e16c6c0, 0xc42023dfb0)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:852 +0x30c
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4203d3b80, 0xc4203d3b80, 0x55726e473180)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:800 +0x2d
main.main()
	/go/src/github.com/docker/cli/cmd/docker/docker.go:180 +0xde
```

**Describe the results you expected:**

The memory reservation being applied to the service.

**Output of `docker version`:**

```
Client:
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        4d60db4
 Built:             Wed Nov  7 00:49:01 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:16:44 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 3
 Paused: 0
 Stopped: 1
Images: 4
Server Version: 18.09.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: active
 NodeID: 3stzeriddfxiemidbnihe5tsk
 Is Manager: true
 ClusterID: m55vlytzzhdype3xq7e4a2pp4
 Managers: 4
 Nodes: 4
 Default Address Pool: 10.0.0.0/8  
 SubnetSize: 24
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.31.55.57
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: c4446665cb9c30056f4998ed953e6d4ff22c7c39
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-1021-aws
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.41GiB
Name: ip-172-31-55-57
ID: TBJB:ZCWP:TBTN:3GLR:36X7:KHOP:5LPB:6DUK:PP5T:QBU6:PXIU:VYHR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Our swarm is operating 100% in AWS EC2 instances.

**Comment by thaJeztah** [comment]
> and only one of those makes the issue occur.

Hm, trying to reproduce the issue, but it doesn't reproduce on my machine 😞. Could it be that one of those machines was upgraded from an older version and the other one a fresh install?


I see where this could go wrong though; working on a fix.

**Comment by thaJeztah** [comment]
Pull request: https://github.com/docker/cli/pull/1566

**Comment by jjcorrea** [comment]
Hi @thaJeztah, thanks for the quick answer. I reviewed the environment from bottom up, and the only difference I see between some of the instances which are part of the swarm is in the kernel release minor version. The same scenario happens independently of the node where we run the command. Do you see the possibility of this affecting in this scenario in some way? 

Also, do you see any way to avoid this problem? I tried to send both limits and reservations as parameters to the cli, but the same error occurs

**Comment by jjcorrea** [comment]
Regarding the machines, all of those which are part of the swarm had docker provisioned in a fresh install

**Comment by thaJeztah** [comment]
Wondering if it's related to swap limit not being supported on that machine (I actually don't think it should make a difference, because "reservation" should only update the value for _accounting_, not actually setting limits or constraints on the containers);

```
WARNING: No swap limit support
```

If you run the check-config script on both machines, is there anything that stands out as a difference? https://raw.githubusercontent.com/moby/moby/master/contrib/check-config.sh

The PR I opened should resolve the panic (the panic is in the CLI, so won't result in an outage of services, but is good to account for)

A version of the CLI with the fix should be available in the nightly builds tomorrow / today, so if you have a chance to verify if my PR fixes the issue, that would be appreciated; you can find nightly static builds here; https://download.docker.com/linux/static/nightly/x86_64/

(I don't think today's nightly build has the fix yet, but it should be there tomorrow)

**Comment by jjcorrea** [comment]
Hey @thaJeztah 

Running the check-config script shows all nodes have the exact same config (including required and optional features).

Thats great, we're going to test the nightly build as soon as its available and post the results here

**Comment by thaJeztah** [comment]
> Running the check-config script shows all nodes have the exact same config (including required and optional features).

🤔 no direct clue then 😓 

> Thats great, we're going to test the nightly build as soon as its available

I see new builds are available (`2018-12-14 08:31:13 53.6 MiB`), so I expect them to have the fix 👍

**Comment by jjcorrea** [comment]
Hey @thaJeztah !

We tested the fix from `20181213123650-d79e916` build and the reservation is now applied as expected 🍺🍻

**Comment by thaJeztah** [comment]
Perfect, thanks so much for testing! I'll close this ticket; the fix will be included in the 18.09.1 patch release (hopefully released this week)
